crates/oxc_linter/src/rules/react/jsx_no_useless_fragment.rs

```diff
@@ -0,0 +1,305 @@
+                if !is_jsx_fragment(&jsx_elem.opening_element) {
+                    return;
+                }
+                self.check_element(node, jsx_elem, ctx);
+            }
+            AstKind::JSXFragment(jsx_elem) => {
+                self.check_fragment(node, jsx_elem, ctx);
+            }
+            _ => {}
+        }
+    }
+}
+
+impl JsxNoUselessFragment {
+    fn check_element(&self, node: &AstNode, elem: &JSXElement, ctx: &LintContext) {
+        if jsx_elem_has_key_attr(elem) {
+            return;
+        }
+
+        if has_less_than_two_children(&elem.children)
+            && !is_fragment_with_only_text_and_is_not_child(node.id(), &elem.children, ctx)
+            && !(self.allow_expressions && is_fragment_with_single_expression(&elem.children))
+        {
+            ctx.diagnostic(JsxNoUselessFragmentDiagnostic::NeedsMoreChildren(elem.span));
+        }
+
+        if is_child_of_html_element(node, ctx) {
+            ctx.diagnostic(JsxNoUselessFragmentDiagnostic::ChildOfHtmlElement(elem.span));
+        }
+    }
+    fn check_fragment(&self, node: &AstNode, elem: &JSXFragment, ctx: &LintContext) {
+        if has_less_than_two_children(&elem.children)
+            && !is_fragment_with_only_text_and_is_not_child(node.id(), &elem.children, ctx)
+            && !(self.allow_expressions && is_fragment_with_single_expression(&elem.children))
+        {
+            ctx.diagnostic(JsxNoUselessFragmentDiagnostic::NeedsMoreChildren(elem.span));
+        }
+
+        if is_child_of_html_element(node, ctx) {
+            ctx.diagnostic(JsxNoUselessFragmentDiagnostic::ChildOfHtmlElement(elem.span));
+        }
+    }
+}
+
+fn jsx_elem_has_key_attr(elem: &JSXElement) -> bool {
+    elem.opening_element.attributes.iter().any(|attr| {
+        let JSXAttributeItem::Attribute(attr) = attr else { return false };
+
+        let JSXAttributeName::Identifier(ident) = &attr.name else { return false };
+
+        ident.name == "key"
+    })
+}
+
+fn is_fragment_with_single_expression(children: &oxc_allocator::Vec<'_, JSXChild<'_>>) -> bool {
+    let children = children.iter().filter(|v| is_padding_spaces(v)).collect::<Vec<_>>();
+
+    children.len() == 1 && matches!(children[0], JSXChild::ExpressionContainer(_))
+}
+
+fn is_padding_spaces(v: &JSXChild<'_>) -> bool {
+    if let JSXChild::Text(v) = v {
+        return !(v.value.trim().is_empty() && v.value.contains('\n'));
+    }
+
+    true
+}
+
+fn is_child_of_html_element(node: &AstNode, ctx: &LintContext) -> bool {
+    if let Some(AstKind::JSXElement(elem)) = ctx.nodes().parent_kind(node.id()) {
+        if is_html_element(&elem.opening_element.name) {
+            return true;
+        }
+    }
+
+    false
+}
+
+fn is_html_element(elem_name: &JSXElementName) -> bool {
+    let JSXElementName::Identifier(ident) = elem_name else { return false };
+
+    lazy_static! {
+        static ref HTML_ELEM_PAT: Regex = Regex::new(r"^[a-z]+$").unwrap();
```

Comment:
This seems like a simple `.starts_with` + check for lowercase, no regex should be required.

Reply:
I try to avoid regex as much as possible due to its drawbacks such as performance.

Reply:
nice, didn't know you could do that :slightly_smiling_face: